File: go/x402/adapters/x402_adapter.go
package adapters

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"sync"

	"github.com/okx/payments/go/paymentrouter"
	x402http "github.com/okx/payments/go/x402/http"
)

// Ensure X402Adapter implements paymentrouter.ProtocolAdapter.
var _ paymentrouter.ProtocolAdapter = (*X402Adapter)(nil)

// X402Adapter implements paymentrouter.ProtocolAdapter using the x402 HTTP server.
type X402Adapter struct {
	server           *x402http.HTTPServer
	priority         int
	registeredRoutes map[string]bool // tracks routes already added via AddRoutes
	initOnce         sync.Once
	initErr          error
}

// NewX402Adapter creates a new X402Adapter. The server can be created with nil
// routes — routes are registered lazily from the paymentrouter route config JSON.
func NewX402Adapter(server *x402http.HTTPServer) *X402Adapter {
	return &X402Adapter{server: server, priority: 20, registeredRoutes: make(map[string]bool)}
}

// initialize calls the underlying x402 server's Initialize method once to
// populate the facilitator client map from the /supported endpoint.
func (a *X402Adapter) initialize(ctx context.Context) error {
	a.initOnce.Do(func() {
		a.initErr = a.server.Initialize(ctx)
	})
	return a.initErr
}

// ensureRoute registers the route with the x402 server if not already registered.
func (a *X402Adapter) ensureRoute(method, path string, cfg any) {
	endpoint := fmt.Sprintf("%s %s", method, path)
	if a.registeredRoutes[endpoint] {
		return
	}
	rc, ok := cfg.(x402http.RouteConfig)
	if !ok {
		return
	}
	a.server.AddRoutes(x402http.RoutesConfig{endpoint: rc})
	a.registeredRoutes[endpoint] = true
}

// Name returns the protocol name.
func (a *X402Adapter) Name() string { return "x402" }

// Priority returns the adapter's dispatch priority.
func (a *X402Adapter) Priority() int { return a.priority }

// Detect returns true if the request carries an x402 payment header.
func (a *X402Adapter) Detect(r *http.Request) bool {
	return r.Header.Get("PAYMENT-SIGNATURE") != "" || r.Header.Get("X-Payment") != ""
}

// GetChallenge builds a payment-required (402) challenge by delegating to the x402 server.
// It returns the response headers that should be forwarded to the caller.
func (a *X402Adapter) GetChallenge(ctx context.Context, r *http.Request, cfg any) (http.Header, error) {
	if err := a.initialize(ctx); err != nil {
		return nil, fmt.Errorf("x402: initialize: %w", err)
	}
	a.ensureRoute(r.Method, r.URL.Path, cfg)

	reqCtx := x402http.HTTPRequestContext{
		Adapter: &netHTTPAdapter{r: r},
		Path:    r.URL.Path,
		Method:  r.Method,
	}

	result := a.server.ProcessHTTPRequest(ctx, reqCtx, nil)
	if result.Type == x402http.ResultPaymentError && result.Response != nil {
		headers := make(http.Header)
		for k, v := range result.Response.Headers {
			headers.Set(k, v)
		}
		return headers, nil
	}

	return http.Header{}, nil
}

// Handle processes an incoming payment request.
// On verified payment it copies settlement headers to w and returns nil.
// On no-payment-required it returns nil without touching w.
// On any error it writes an error response to w and returns a non-nil error.
func (a *X402Adapter) Handle(w http.ResponseWriter, r *http.Request, cfg any) error {
	if err := a.initialize(r.Context()); err != nil {
		return fmt.Errorf("x402: initialize: %w", err)
	}
	a.ensureRoute(r.Method, r.URL.Path, cfg)

	reqCtx := x402http.HTTPRequestContext{
		Adapter:       &netHTTPAdapter{r: r},
		Path:          r.URL.Path,
		Method:        r.Method,
		PaymentHeader: r.Header.Get("PAYMENT-SIGNATURE"),
	}

	result := a.server.ProcessHTTPRequest(r.Context(), reqCtx, nil)

	switch result.Type {
	case x402http.ResultPaymentVerified:
		if result.Response != nil {
			for k, v := range result.Response.Headers {
				w.Header().Set(k, v)
			}
		}
		// Settle payment after successful verification
		if result.PaymentPayload != nil && result.PaymentRequirements != nil {
			settleResult := a.server.ProcessSettlement(r.Context(), *result.PaymentPayload, *result.PaymentRequirements, nil)
			if settleResult != nil {
				if !settleResult.Success {
					if settleResult.Response != nil {
						for k, v := range settleResult.Response.Headers {
							w.Header().Set(k, v)
						}
						w.WriteHeader(settleResult.Response.Status)
						if settleResult.Response.Body != nil {
							body, _ := json.Marshal(settleResult.Response.Body)
							w.Write(body) //nolint:errcheck
						}
					}
					return fmt.Errorf("x402: settlement failed: %s", settleResult.ErrorReason)
				}
				for k, v := range settleResult.Headers {
					w.Header().Set(k, v)
				}
			}
		}
		return nil

	case x402http.ResultNoPaymentRequired:
		return nil

	default:
		// ResultPaymentError or unexpected result
		var bodyStr string
		if result.Response != nil {
			for k, v := range result.Response.Headers {
				w.Header().Set(k, v)
			}
			w.WriteHeader(result.Response.Status)
			if result.Response.Body != nil {
				body, err := json.Marshal(result.Response.Body)
				if err == nil {
					bodyStr = string(body)
					_, _ = w.Write(body)
				}
			}
		} else {
			w.WriteHeader(http.StatusPaymentRequired)
		}
		return fmt.Errorf("x402: payment error (result type: %s, body: %s)", result.Type, bodyStr)
	}
}

// ============================================================================
// netHTTPAdapter — adapts *http.Request to x402http.HTTPAdapter
// ============================================================================

type netHTTPAdapter struct {
	r *http.Request
}

func (a *netHTTPAdapter) GetHeader(name string) string { return a.r.Header.Get(name) }
func (a *netHTTPAdapter) GetMethod() string            { return a.r.Method }
func (a *netHTTPAdapter) GetPath() string              { return a.r.URL.Path }

func (a *netHTTPAdapter) GetURL() string {
	scheme := "http"
	if a.r.TLS != nil || a.r.Header.Get("X-Forwarded-Proto") == "https" {
		scheme = "https"
	}
	host := a.r.Host
	if host == "" {
		host = a.r.URL.Host
	}
	return scheme + "://" + host + a.r.RequestURI
}

func (a *netHTTPAdapter) GetAcceptHeader() string { return a.r.Header.Get("Accept") }
func (a *netHTTPAdapter) GetUserAgent() string    { return a.r.Header.Get("User-Agent") }

File: python/x402/src/x402/adapters/x402_adapter.py
"""PaymentRouter ProtocolAdapter for x402.

Mirrors go/x402/adapters/x402_adapter.go.
Wraps an x402HTTPResourceServer and delegates detect/challenge/handle.
"""

from __future__ import annotations

from typing import Any

from x402.http.x402_http_server import x402HTTPResourceServer
from x402.http.constants import PAYMENT_SIGNATURE_HEADER
from x402.http.types import (
    HTTPAdapter,
    HTTPRequestContext,
    RouteConfig,
    RESULT_NO_PAYMENT_REQUIRED,
    RESULT_PAYMENT_VERIFIED,
)


class _RequestAdapterBridge:
    """Bridges paymentrouter's RequestAdapter to x402's HTTPAdapter protocol."""

    def __init__(self, request: Any) -> None:
        self._request = request

    def get_header(self, name: str) -> str:
        return self._request.get_header(name)

    def get_method(self) -> str:
        return self._request.method

    def get_path(self) -> str:
        return self._request.path

    def get_url(self) -> str:
        return self._request.url

    def get_accept_header(self) -> str:
        return self._request.get_header("accept")

    def get_user_agent(self) -> str:
        return self._request.get_header("user-agent")


class X402Adapter:
    """PaymentRouter ProtocolAdapter for x402 protocol.

    Wraps an x402HTTPResourceServer. Detects requests with
    PAYMENT-SIGNATURE or X-Payment header.
    """

    def __init__(self, server: x402HTTPResourceServer, priority: int = 20) -> None:
        self._server = server
        self._priority = priority
        self._registered_routes: set[str] = set()

    @property
    def name(self) -> str:
        return "x402"

    @property
    def priority(self) -> int:
        return self._priority

    def detect(self, request: Any) -> bool:
        return bool(
            request.get_header(PAYMENT_SIGNATURE_HEADER)
            or request.get_header(PAYMENT_SIGNATURE_HEADER.lower())
            or request.get_header("x-payment")
        )

    def _maybe_initialize(self) -> None:
        if not getattr(self, "_initialized", False):
            self._server._server.initialize()
            self._initialized = True

    def _ensure_route(self, method: str | Any, path: str | Any, cfg: Any) -> None:
        endpoint = f"{method} {path}"
        if endpoint in self._registered_routes:
            return
        if isinstance(cfg, RouteConfig):
            self._server.add_routes({endpoint: cfg})
            self._registered_routes.add(endpoint)

    async def get_challenge(self, request: Any, cfg: Any) -> dict[str, str]:
        """Generate 402 challenge.

        Tries the x402 server first. Falls back to building the header
        directly from RouteConfig if the server can't resolve options.
        """
        self._ensure_route(request.method, request.path, cfg)
        self._maybe_initialize()

        # Try server-generated challenge first
        bridge = _RequestAdapterBridge(request)
        context = HTTPRequestContext(
            adapter=bridge,
            path=request.path,
            method=request.method,
        )

        result = await self._server.process_http_request(context)
        if result.response is not None and result.response.headers:
            return dict(result.response.headers)

        # Fallback: build challenge directly from RouteConfig
        if not isinstance(cfg, RouteConfig):
            return {}
        return self._build_challenge_from_config(cfg, request)

    def _build_challenge_from_config(self, cfg: RouteConfig, request: Any) -> dict[str, str]:
        """Build PAYMENT-REQUIRED header from RouteConfig as fallback."""
        from x402.http.utils import encode_payment_required_header
        from x402.schemas import PaymentRequired, PaymentRequirements

        requirements_list = []
        for opt in cfg.accepts if isinstance(cfg.accepts, list) else [cfg.accepts]:
            price_str = str(opt.price)
            if price_str.startswith("$"):
                price_str = price_str[1:]
            req = PaymentRequirements(
                scheme=opt.scheme,
                network=opt.network,
                asset=opt.extra.get("asset", "") if opt.extra else "",
                amount=price_str,
                pay_to=opt.pay_to if isinstance(opt.pay_to, str) else "",
                max_timeout_seconds=opt.max_timeout_seconds or 300,
                extra=opt.extra or {},
            )
            requirements_list.append(req)

        if not requirements_list:
            return {}

        payment_required = PaymentRequired(accepts=requirements_list)
        header_value = encode_payment_required_header(payment_required)
        return {"PAYMENT-REQUIRED": header_value}

    async def handle(self, request: Any, cfg: Any) -> Any:
        self._ensure_route(request.method, request.path, cfg)
        self._maybe_initialize()

        bridge = _RequestAdapterBridge(request)
        context = HTTPRequestContext(
            adapter=bridge,
            path=request.path,
            method=request.method,
        )

        result = await self._server.process_http_request(context)

        if result.type == RESULT_PAYMENT_VERIFIED:
            # Process settlement
            if result.payment_payload is not None and result.payment_requirements is not None:
                settle_result = await self._server.process_settlement(
                    result.payment_payload, result.payment_requirements, context
                )
                if settle_result and not settle_result.success:
                    raise ValueError(f"x402: settlement failed: {settle_result.error_reason}")
                return {"settled": True, "headers": settle_result.headers if settle_result else {}}
            return {"verified": True}

        if result.type == RESULT_NO_PAYMENT_REQUIRED:
            return {"no_payment_required": True}

        # Payment error — raise so paymentrouter returns 402
        raise ValueError(f"x402: payment error")
